### PR TITLE
XW-3861 | Reducing the amount of LiftIgniter API requests

### PR DESCRIPTION
**Pull request by JamesHawking**
[Ticket](https://wikia-inc.atlassian.net/browse/XW-3861)

@Wikia/x-wing

### CODE REVIEW

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 1]
```diff
@@ -86,7 +87,20 @@ require([
 	// fetch data for all recirculation modules
 	// TODO lazy load some data on scroll
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 2]
```diff
@@ -86,7 +87,20 @@ require([
...
+
+		if (aproachingMCF && !wikiArticlesFetched) {
+			liftigniter.fetch('wiki');
+			wikiArticlesFetched = true;
+		}
+	});
+
 	discussions.fetch();
 
 	// TODO handle errors
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [recirculation.js#L107](https://github.com/Wikia/app/blob/59d5fa75534a290db9d3254077d077ef164d41c8/extensions/wikia/Recirculation/js/recirculation.js#L107): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [recirculation.js#L106](https://github.com/Wikia/app/blob/8c75e2d4799596ba5b06be1fec72a694bc124d0a/extensions/wikia/Recirculation/js/recirculation.js#L106): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 3]
```diff
@@ -84,9 +85,21 @@ require([
...
 
 	// fetch data for all recirculation modules
-	// TODO lazy load some data on scroll
 	liftigniter.fetch('ns');
-	liftigniter.fetch('wiki');
+
+	window.addEventListener('scroll', function () {
+		var mcfOffset = $mixedContentFooter.offset().top,
+			scrollPosition = $(window).scrollTop(),
+			windowInnerHeight = $(window).innerHeight(),
```
Comment (rogatty) [inline extensions/wikia/Recirculation/js/recirculation.js]: This is pretty heavy to run on every scroll event (there's a lot of them). Consider using a throttle.

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 4]
```diff
@@ -84,10 +85,24 @@ require([
...
+			if (aproachingMCF) {
+				liftigniter.fetch('wiki');
+				discussions.fetch();
+				wikiArticlesFetched = true;
+			}
+		}
+	});
+
 
 	// TODO handle errors
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 5]
```diff
@@ -84,10 +85,24 @@ require([
...
 	// TODO LI tracking
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Review by latata — APPROVED** [review]

**Comment by latata** [comment]
I'd add `display:none` to mixed content footer and show it on data ready

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 6]
```diff
@@ -84,10 +87,24 @@ require([
 	});
 
 	// fetch data for all recirculation modules
-	// TODO lazy load some data on scroll
 	liftigniter.fetch('ns');
-	liftigniter.fetch('wiki');
-	discussions.fetch();
+
+	window.addEventListener('scroll', $.throttle(50, function () {
+		if(!wikiArticlesFetched) {
```
Comment (rogatty) [inline extensions/wikia/Recirculation/js/recirculation.js]: How about removing the event listener when `wikiArticlesFetched === true`?

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 7]
```diff
@@ -84,10 +87,27 @@ require([
...
+					discussions.fetch();
+					wikiArticlesFetched = true;
+					window.removeEventListener('scroll', lazyLoadHandler);
+				}
+			}
+		}))()
+	});
+
 
 	// TODO handle errors
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 8]
```diff
@@ -84,10 +87,27 @@ require([
...
 	// TODO LI tracking
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 9]
```diff
@@ -84,10 +87,27 @@ require([
 	});
 
 	// fetch data for all recirculation modules
-	// TODO lazy load some data on scroll
 	liftigniter.fetch('ns');
-	liftigniter.fetch('wiki');
-	discussions.fetch();
+
+	window.addEventListener('scroll', function lazyLoadHandler() {
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js]: MAJOR Make this function anonymous by removing its name: 'function() {...}'. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3ANamedFunctionExpression)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 10]
```diff
@@ -20,6 +20,8 @@ require([
 	'use strict';
 
 	var $mixedContentFooter = $('#mixed-content-footer'),
+		$mixedContentFooterContent = $('.mcf-content'),
+	    wikiArticlesFetched = false,
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js]: MAJOR Remove the declaration of the unused 'wikiArticlesFetched' variable. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AUnusedVariable)

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 11]
```diff
@@ -84,10 +87,25 @@ require([
...
+				discussions.fetch();
+				window.removeEventListener('scroll', lazyLoadHandler);
+
+			}
+	});
+
+	window.addEventListener('scroll', lazyLoadHandler);
+
 
 	// TODO handle errors
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js:109]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 12]
```diff
@@ -84,10 +87,25 @@ require([
...
 	// TODO LI tracking
```
Comment (WikiaTech) [inline extensions/wikia/Recirculation/js/recirculation.js:110]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Pull request merged**